kernel/src/actions/mod.rs

```diff
@@ -154,6 +162,59 @@ impl Protocol {
     }
 }
 
+mod protocol {
+    use super::*;
+
+    #[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+    pub(crate) struct UnvalidatedProtocol(Protocol);
+
+    impl UnvalidatedProtocol {
+        #[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+        pub(crate) fn validate(self) -> DeltaResult<Protocol> {
+            let protocol = self.0;
+            require!(
+                protocol.min_reader_version <= KERNEL_READER_VERSION,
+                Error::invalid_protocol(format!(
+                    "Minimum reader version exceeds kernel reader version {KERNEL_READER_VERSION}"
+                ))
+            );
+            require!(
+                protocol.min_writer_version <= KERNEL_WRITER_VERSION,
+                Error::invalid_protocol(format!(
+                    "Minimum writer version exceeds kernel writer version {KERNEL_WRITER_VERSION}"
+                ))
+            );
+            if protocol.min_reader_version == 3 {
+                require!(
+                    protocol.reader_features.is_some(),
+                    Error::invalid_protocol(
+                        "Reader features must be present when minimum reader version = 3"
+                    )
+                );
+            }
+            if protocol.min_writer_version == 7 {
+                require!(
+                    protocol.writer_features.is_some(),
+                    Error::invalid_protocol(
+                        "Writer features must be present when minimum writer version = 7"
+                    )
+                );
+            }
+            require!(
+                !protocol.has_reader_feature(&ReaderFeatures::V2Checkpoint),
+                Error::unsupported("V2 Checkpoint reader feature is not yet supported")
+            );
```

Comment:
ryan: need an allowlist

Reply:
need RWfeatures to allow unknown case so we can 'catch' unrecognized table features